### ISSUE TITLE
Unexpected "identifier i$0 not defined"

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#185]
Type checking fails with "identifier i$0 not defined" in the following program -- I think that it's a bug.

By commenting the precondition type checking succeeds.

```viper
field value: Int;

define validIndex(a, i) (
   0 <= i && i < |a|
)

define HeapArray(array) (
    forall i: Int :: { array[i] } validIndex(array, i) ==> (array[i] != null && acc(array[i].value))
)

method foo(tree: Ref, leftArray: Seq[Ref], rightArray: Seq[Ref], array: Seq[Ref])
    requires HeapArray(array)
{
    // Type checking fails with "identifier i$0 not defined"
    assert forall i:Int :: { array[i] } validIndex(rightArray, i - |leftArray| - 1) ==> acc(array[i].value)
}
```

[Issue created by fpoli_eth: 2018-05-16]
[Last updated on bitbucket: 2018-12-11]

[Comment created by fabiopakk: 2018-12-11]
This issue was fixed in branch fabiopakk_issue_239_identifier_not_defined, merged into default branch.

**Comment by fpoli** [comment]
> **@alexander_summers** commented on 2017-01-11 15:03

I suggest dropping the `private[ast]` from the case class. It seems to be an attempt to enforce that the consistency checks are made, but those should really be made elsewhere/differently anyway (cf. issue #174)

**Comment by fpoli** [comment]
> **@mschwerhoff** commented on 2017-08-25 14:46

Resolved by @mousam05 with pull request #73
